File: set1/challenge5.hs
-- http://cryptopals.com/sets/1/challenges/5/
import Data.Bits (xor)
import Data.Char (ord)
import Text.Printf (printf)

int2hex :: Int -> [Char]
int2hex i = printf "%02x" i

hexEncode :: [Int] -> [Char]
hexEncode input = concat [int2hex i | i <- input]

encrypt :: [Char] -> [Char] -> [Char]
encrypt input key = hexEncode $ zipWith xor (map ord input) (cycle (map ord key))

main =
    putStrLn $ encrypt "Burning 'em, if you ain't quick and nimble\nI go crazy when I hear a cymbal" "ICE"

File: set1/challenge5.py
# http://cryptopals.com/sets/1/challenges/5/
from itertools import cycle, imap, izip, starmap
from operator import xor


def int2hex(integer):
    return '%0.2x' % integer


def hexencode(integers):
    hexdigits = imap(int2hex, integers)
    return ''.join(hexdigits)


def encrypt(plaintext, key):
    pairs = izip(imap(ord, plaintext), cycle(imap(ord, key)))
    cyphertext = starmap(xor, pairs)
    return hexencode(cyphertext)


if __name__ == '__main__':
    input = "Burning 'em, if you ain't quick and nimble\nI go crazy when I hear a cymbal"
    key = "ICE"
    print(encrypt(input, key))
